classes/Hook/HookDisplayBeforeBodyClosingTag.php

```diff
@@ -49,100 +44,236 @@ public function __construct(Ps_Googleanalytics $module, Context $context)
-                } elseif (is_array($gacart)) {
-                    $it = new RecursiveIteratorIterator(new RecursiveArrayIterator($gacart));
-                    foreach ($it as $v) {
-                        $gaScripts .= $v;
-                    }
-                } else {
-                    $gaScripts .= $gacart;
-                }
-            }
-
-            $ganalyticsDataHandler->manageData('', 'D');
-        }
+        // Log information about product listing
+        $this->saveInformationAboutListing();
+
+        // Flush events stored in data storage from previous pages
+        $this->outputStoredEvents();
+
+        // Add events
+        $this->renderProductListing();
+        $this->renderSearch();
+        $this->renderCartPage();
+        $this->renderBeginCheckout();
+        $this->renderLogin();
+        $this->renderRegistration();
+
+        // Output everything
+        return $gaTagHandler->generate($this->gaScripts);
+    }
 
+    /**
+     * This method renders tracking code for product listings, like category pages.
+     */
+    private function renderProductListing()
+    {
+        // Try to get product list variable
         $listing = $this->context->smarty->getTemplateVars('listing');
+        if (empty($listing['products'])) {
+            return;
+        }
+
+        // Prepare items to our format
         $productWrapper = new ProductWrapper($this->context);
-        $products = $productWrapper->wrapProductList(isset($listing['products']) ? $listing['products'] : []);
+        $items = $productWrapper->prepareItemListFromProductList($listing['products']);
+
+        // Prepare info about the list
+        $item_list_id = $this->context->controller->php_self;
+        $item_list_name = $listing['label'];
 
-        if ($controller_name == 'order' || $controller_name == 'orderopc') {
+        // Render the event
+        $eventData = [
+            'item_list_id' => $item_list_id,
+            'item_list_name' => $item_list_name,
+            'items' => $items,
+        ];
+        $this->gaScripts .= $this->module->getTools()->renderEvent(
+            'view_item_list',
+            $eventData
+        );
+
+        // Render quickview events
+        foreach ($items as $item) {
             $eventData = [
-                'currency' => $this->context->currency->iso_code,
+                'item_list_id' => $item_list_id,
+                'item_list_name' => $item_list_name,
+                'items' => [$item],
             ];
-            $gaScripts = $this->module->getTools()->renderEvent(
-                'begin_checkout',
-                $eventData
+
+            // Keep only product ID if id_product_attribute was appended
+            $productId = explode('-', $item['item_id']);
+            $productId = $productId[0];
+
+            // Render the event wrapped in onclick
+            $this->gaScripts .= '
+            $(\'article[data-id-product="' . $productId . '"] a.quick-view\').on(
```

Comment:
🤮 
Brrrrr 😰 

That's why I suggest to better wrap the string computation part 😅